[Ek Wei Rui] iP

ekweirui opened the pull request:
# **Sky**
Sky frees your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- ELECTRIC FAST to use

All you need to do is,

1. download it from [here](https://github.com/ekweirui/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

## Features

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

## Commands
Some basic commands to start with are:

```
// Ends the program
bye

// Lists all the tasks on your plate
list

// Marks selected task as completed
mark

// Marks selected task as uncompleted
unmark

// Adds a to-do task
todo

// Adds a deadline task
deadline

// Adds an event task
event

// Deletes a task from the list given a task number
delete

// Find a task from the list via keyword
find
```
## 
If you are a java programmer, here's the `main` method to run it:

```java
public static void main(String[] args) {
    new Sky("data/sky.txt").run();
}
```

Commit:
Add Gradle support

ekweirui committed:
Add Sky.java

ekweirui committed:
Add list functionality

ekweirui committed:
Add mark as done functionality

ekweirui committed:
Add ToDos, Deadlines, and Events

ekweirui committed:
Semi-automate the testing of Sky

ekweirui committed:
Handle some errors

ekweirui committed:
Add delete functionality

ekweirui committed:
Add save to/read from hard disk functionality

ekweirui committed:
Add capability to understand date and time

ekweirui committed:
Merge branch 'branch-Level-7'

ekweirui committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Sky.java

ekweirui committed:
Use more OOP

ekweirui committed:
Divide classes into packages

ekweirui committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

ekweirui committed:
Add JUnit tests

ekweirui committed:
Add JavaDoc comments

ekweirui committed:
Add the ability to find a task by searching for a keyword

ekweirui committed:
Follow the coding standard

ekweirui committed:
Merge branch 'branch-A-JavaDoc'

ekweirui committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/sky/Sky.java

ekweirui committed:
Merge branch 'branch-Level-9'

hanwenlai commented on src/main/java/sky/Parser.java:
```diff
@@ -0,0 +1,38 @@
+package sky;
+
+import sky.command.*;
```
Perhaps list the import classes explicitly?

ekweirui replied:
Yep nice spot!

hanwenlai commented on src/main/java/sky/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+     *
+     * @return Size of the task list.
+     */
+    public int size() {
```
Maybe this method should be named getSize()?

ekweirui replied:
Yes I agree. Thanks for the heads up.

hanwenlai commented on src/main/java/sky/Sky.java:
```diff
@@ -0,0 +1,52 @@
+    private Ui ui;
+    private Storage storage;
+
+    public Sky(String filePath) {
```
Perhaps include a header comment for this constructor?

ekweirui replied:
Nice suggestion! Will add it in the future.

hanwenlai commented on src/main/java/sky/exception/TextNoMeaningException.java:
```diff
@@ -0,0 +1,13 @@
+package sky.exception;
+
+public class TextNoMeaningException extends Exception {
```
Perhaps add a header comment for this class?

ekweirui replied:
Good suggestion!

hanwenlai reviewed (commented):
Well done. 👍 Coding standards are obeyed in most of the code.

ekweirui committed:
Replace text-based UI with JavaFX GUI

ekweirui committed:
Merge branch 'branch-Level-10'

elizabethhky commented on src/main/java/sky/Parser.java:
```diff
@@ -0,0 +1,48 @@
+            return new FindCommand(fullCommand);
+        } else {
+            return new InvalidCommand();
+        }
```
I really like the logic applied here in the parser!

ekweirui replied:
Thank you :)

elizabethhky commented on src/main/java/sky/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+ * The TaskList class contains the task list e.g., it has operations to add/delete tasks in the list.
+ */
+public class TaskList {
+    private List<Task> taskList;
+
+    public TaskList() {
+        this.taskList = new ArrayList<>();
```
I noticed that you used `List<Task>` for your `taskList` but set it to be an `ArrayList<>()`. Any reason for why you chose to do it this way instead of just assigning an `ArrayList<Task>` type to your `taskList`?

ekweirui replied:
I did so as I wanted to lessen the burden on the reader by abstracting a tad further about the kind of list being used. However, you may be right that it might be redundant. 🤔

elizabethhky commented on src/main/java/sky/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+    public String printTasks() {
+        String s = "";
+        for (int i = 0; i < this.taskList.size(); i++) {
+            s += "  " + (i + 1) + "." + this.taskList.get(i) + (i == this.taskList.size() - 1 ? "" : "\n");
```
I really like the way you used to decide whether to add a newline! Very neat and clever 👍

ekweirui committed:
Update GUI to colour dialog boxes

ekweirui committed:
Add and use checkstyle to detect coding violations

ekweirui committed:
Update various UI responses

ekweirui committed:
Add assertions to detect programming errors

Currently, there are no assertions in the code base to detect
programming errors.

This needs to change as assertions can help to detect programming
errors early and prevent the chat bot from performing undesirably.

Add assertions to detect programming errors.

Assertions are not only simple, but their impact on performance is considered low and worth the additional safety they provide.

ekweirui committed:
Provide a way to undo commands

Currently, there is no way to undo a command.

This needs to change as users might have misspelt their tasks or
accidentally deleted a task.

Provide a way to undo commands by using an undo command, followed by
specifying the number of steps to revert by.

This is done so that users can have the flexibility to revert their
state by multiple increments.

ekweirui committed:
Improve code quality

Currently, there are a few functions who do not follow the principle of
single level of abstraction SLAP).

This needs to change as otherwise, it can lead to unclean code and be
messy.

Improve code quality by implementing SLAP on the code.

SLAP helps to ensure that abstraction levels are kept consistent
throughout a function and makes code easily understood.

ekweirui committed:
Merge pull request #1 from ekweirui/branch-A-Assertions

Add assertions to detect programming errors

ekweirui committed:
Merge branch 'master' into branch-C-Undo

ekweirui committed:
Merge pull request #2 from ekweirui/branch-C-Undo

Provide a way to undo commands

ekweirui committed:
Merge branch 'master' of https://github.com/ekweirui/ip

# Conflicts:
#	src/main/java/sky/command/DeadlineCommand.java
#	src/main/java/sky/command/DeleteCommand.java
#	src/main/java/sky/command/EventCommand.java
#	src/main/java/sky/command/MarkCommand.java
#	src/main/java/sky/command/UnmarkCommand.java

ekweirui committed:
Merge branch 'master' into branch-A-CodeQuality

ekweirui committed:
Merge pull request #3 from ekweirui/branch-A-CodeQuality

Improve code quality

ekweirui committed:
Merge branch 'master' of https://github.com/ekweirui/ip

quannam0124 reviewed (commented):
Overall, your code is very easy to read. Good job!

elizabethhky commented on src/main/java/sky/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,107 @@
+        return false;
+    }
+
+    private String produceDateAndTimeForDeadline(String s) throws TextNoMeaningException {
```
Should there be a JavaDoc comment for this method?

ekweirui replied:
I agree. Although JavaDocs is not a necessity for private methods, it's a good thing to have them for all methods. Good point!

elizabethhky commented on src/main/java/sky/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,107 @@
+            LocalDate d1 = LocalDate.parse(dateGiven);
+            String dateString = d1.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+
+            // If time is provided
+            if (arrOfStrings.length == 2) {
```
I noticed that there are comments in your code that explains **how** the code works. Perhaps you can consider ways to abstract these out into another method to apply SLAP or avoid magic literals by abstracting out these variables? 
```suggestion
            boolean isTimeProvided = arrOfStrings.length == 2
```

ekweirui replied:
Thanks for the suggestion! It's a good one and I am working on it now 😄

elizabethhky commented on src/main/java/sky/command/DeleteCommand.java:
```diff
@@ -0,0 +1,56 @@
+    public String execute(TaskList taskList, Storage storage, History history)
+            throws TextNoMeaningException, IOException {
+        try {
+            int taskNum = generateTaskNumToDelete();
+            Task task = taskList.getTask(taskNum);
+            taskList.removeTask(task);
+            storage.reWriteDataFile(taskList);
+            history.addHistoryInTime(taskList);
+            return generateResponse(task, taskList);
```
I like the way you organised the code into different levels of abstraction for better readability! 👍

elizabethhky commented on src/main/java/sky/command/EventCommand.java:
```diff
@@ -0,0 +1,108 @@
+
+    private String produceDateAndTimeForEvent(String s) throws TextNoMeaningException {
+        try {
+            String[] arrOfStrings = s.split(" ");
+            if (arrOfStrings.length != 2) {
+                throw new TextNoMeaningException("Provide the date and time after \"/at\""
+                        + " as: \"yyyy/mm/dd XXXX-XXXX\", where XXXX is time in 24-hours.");
+            }
+            String dateGiven = arrOfStrings[0].replaceAll("/", "-");
+            LocalDate d1 = LocalDate.parse(dateGiven);
+            String dateString = d1.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+
+            String timeGiven1 = arrOfStrings[1].substring(0, 2) + ":" + arrOfStrings[1].substring(2, 4);
+            String timeGiven2 = arrOfStrings[1].substring(5, 7) + ":" + arrOfStrings[1].substring(7);
+            LocalTime t1 = LocalTime.parse(timeGiven1);
+            LocalTime t2 = LocalTime.parse(timeGiven2);
+            String timeString1 = t1.format(DateTimeFormatter.ofPattern("h:mma"));
+            String timeString2 = t2.format(DateTimeFormatter.ofPattern("h:mma"));
+            String finalTimeString = timeString1 + "-" + timeString2;
+
+            return dateString + ", " + finalTimeString;
```
Is there a way to shorten this code?

ekweirui replied:
You are right, there are plenty of ways and reasons to do so. Thanks for pointing it out! Will work on it.

elizabethhky commented on src/main/java/sky/storage/Storage.java:
```diff
@@ -0,0 +1,136 @@
+        if (!file.exists()) {
+            return new ArrayList<>();
+        }
+        try {
+            List<Task> taskList = new ArrayList<>();
+            Scanner scanner = new Scanner(file);
+            while (scanner.hasNext()) {
+                String textLine = scanner.nextLine();
+                char taskType = textLine.charAt(1);
+                boolean isMarked = textLine.charAt(4) == 'X';
+                String description = textLine.substring(7);
+
+                if (taskType == 'T') {
+                    Task task = new ToDo(description);
+                    taskList.add(task);
+                } else if (taskType == 'D') {
+                    int indexOfBracket = -1;
+                    for (int i = 0; i < description.length(); i++) {
+                        if (description.charAt(i) == '(') {
+                            indexOfBracket = i;
+                        }
+                    }
+                    String taskDescription = description.substring(0, indexOfBracket - 1);
+                    String taskBy = description.substring(indexOfBracket + 5, description.length() - 1);
+                    Task task = new Deadline(taskDescription, taskBy);
+                    taskList.add(task);
+                } else if (taskType == 'E') {
+                    int indexOfBracket = -1;
+                    for (int i = 0; i < description.length(); i++) {
+                        if (description.charAt(i) == '(') {
+                            indexOfBracket = i;
+                        }
+                    }
```
I noticed that this block of code follows the arrow-head style. Perhaps you can try to change your code to avoid this for better readability? e.g using guard clauses for happy paths

ekweirui replied:
You are right that this code could be too nested. Thanks for pointing this out :)

elizabethhky commented on src/test/java/sky/command/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Test
+    public void deadlineTestOnEmptyList() {
```
I noticed that your test methods do not follow the recommended format of `featureUnderTest_testScenario_expectedBehavior()`. Perhaps you can consider changing your test methods to this naming convention for easier reference? Otherwise, I think it should be fine to leave them as they are as well 😊

elizabethhky commented on src/test/java/sky/command/TodoCommandTest.java:
```diff
@@ -0,0 +1,42 @@
+    private History history = new History(this.sky, this.taskList);
+
+    @Test
+    public void dummyTest() {
+        assertEquals(2, 2);
+    }
```
Perhaps consider deleting this dummyTest in your actual code?

elizabethhky reviewed (commented):
Overall, good job on your code! I like the way some of your code is structured. Do keep a look out for the readability of your code and some missing JavaDoc comments 😊 Well done!

ekweirui committed:
Improve the GUI

ekweirui committed:
Add title to stage

ekweirui committed:
Fix return message in mark

ekweirui committed:
Improve code quality by applying SLAP

ekweirui committed:
Add a user guide

ekweirui committed:
Update user guide
